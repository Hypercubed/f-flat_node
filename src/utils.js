import {typed, Action} from './types/index';
import {inspect} from 'util';

/* export function nAry (n, fn) {
  switch (n) {
    case 0: return function () { return fn.call(this); };
    case 1: return function (a0) { return fn.call(this, a0); };
    case 2: return function (a0, a1) { return fn.call(this, a0, a1); };
    case 3: return function (a0, a1, a2) { return fn.call(this, a0, a1, a2); };
    case 4: return function (a0, a1, a2, a3) { return fn.call(this, a0, a1, a2, a3); };
    case 5: return function (a0, a1, a2, a3, a4) { return fn.call(this, a0, a1, a2, a3, a4); };
    case 6: return function (a0, a1, a2, a3, a4, a5) { return fn.call(this, a0, a1, a2, a3, a4, a5); };
    case 7: return function (a0, a1, a2, a3, a4, a5, a6) { return fn.call(this, a0, a1, a2, a3, a4, a5, a6); };
    case 8: return function (a0, a1, a2, a3, a4, a5, a6, a7) { return fn.call(this, a0, a1, a2, a3, a4, a5, a6, a7); };
    case 9: return function (a0, a1, a2, a3, a4, a5, a6, a7, a8) { return fn.call(this, a0, a1, a2, a3, a4, a5, a6, a7, a8); };
    case 10: return function (a0, a1, a2, a3, a4, a5, a6, a7, a8, a9) { return fn.call(this, a0, a1, a2, a3, a4, a5, a6, a7, a8, a9); };
    default: throw new Error('First argument to nAry must be a non-negative integer no greater than ten');
  }
}

export function unary (fn) {
  return nAry(1, fn);
} */

export const arrayRepeat = (a, n) => {
  n = Number(n) | 0;
  if (n < 1) {
    return [];
  }
  let r = [];
  for (let i = 0; i < n; i++) {
    r = r.concat(a);
  }
  return r;
};

export const arrayMul = (lhs, rhs) => {
  return lhs.reduce((p, n) => {
    return p.concat([n], rhs);
  }, []);
};

export function pluck (context, path) {
  // if (path === '.') return undefined;

  const pathParts = path.split('.');
  const ii = pathParts.length;

  for (let i = 0; i < ii; i++) {
    if (!context) {
      break;
    }
    context = context[pathParts[i]];
  }

  return context;
}

/* export function copy (src) {
  if (src instanceof Array) {
    return src.slice();
  }
  return src;
} */

/* istanbul ignore next */
export function noop () {}

/* istanbul ignore next */
export function throwError (e) {
  throw e;
}

/* export function isLineTerminator (ch) {
  return (ch === '\n') || (ch === '\r') || (ch === 0x2028) || (ch === 0x2029);
}

export function isWhitespace (ch) {
  return (ch === ',') || // camma
        (ch === ' ') ||  // space
        (ch === '\t') || (ch === '\v') || // tab
        isLineTerminator(ch) || // newlines
        (ch === '\n') ||
        (ch === 0xB) ||
        (ch === 0xC) ||
        (ch === 0xA0) ||
        (ch >= 0x1680 && '\u1680\u180E\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u3000\uFEFF'.indexOf(String.fromCharCode(ch)) > 0);
} */

/* export function isWhitespace (ch) {
  return (ch === ' ' || ch === '\r' || ch === '\t' ||
          ch === '\n' || ch === '\v' || ch === '\u00A0' ||
          ch === ',');
} */

/* export function isQuote (ch) {
  return (ch === '"' || ch === '\'');
} */

/* export function isBracket (ch) {
  return (ch === '{' || ch === '}' ||
          ch === '[' || ch === ']' ||
          ch === '(' || ch === ')');
} */

/* export function isNumeric (num) {
  return !isNaN(num);
}

export function isBoolean (string) {
  var lc = string.toLowerCase();
  return lc === 'true' || lc === 'false';
} */

export function isPromise (val) {
  return val && typeof val.then === 'function';
}

export function isDefined (val) {
  return typeof val !== 'undefined';
}

const __eql = typed('eql', {
  'Array, Array': (a, b) => {
    if (a.length !== b.length) {
      return false;
    }
    for (let i = 0; i < a.length; i++) {
      if (!eql(a[i], b[i])) {
        return false;
      }
    }
    return true;
  },
  'Action, Action': (a, b) => {
    return a.value === b.value;
  },
  'BigNumber, BigNumber | number': (a, b) => {
    return a.equals(b);
  },
  'any, any': (a, b) => {
    return a === b;
  }
});

export function eql (a, b) {
  if (a === b || a == b) {  // eslint-disable-line eqeqeq
    return true;
  }
  return __eql(a, b);
}

/* export function eql (a, b) {
  if (a == b) { return true; }
  if (a instanceof Array && b instanceof Array) {
    if (a.length === b.length) {
      for (var i = 0; i < a.length; i++) {
        if (!eql(a[i], b[i])) { return false; }
      }
      return true;
    }
  }
  if (a instanceof Action && b instanceof Action) {
    return a.value === b.value;
  }
  if (a instanceof BigNumber && b instanceof BigNumber) {
    return a.equals(b);
  }
  return is.equal(a, b);
} */

/* export function toLiteral (d) {
  if (isNumeric(d)) {
    return Object.freeze(new BigNumber(d));
  }
  if (isBoolean(d)) {
    return d.toLowerCase() === 'true';
  }
  if (d === 'i') {
    return I;
  }
  return processIdentifier(d);
} */

/* export const asap = typeof setImmediate === 'function' ? setImmediate  // from https://github.com/folktale/data.task/blob/master/lib/task.js#L9
  : typeof process !== 'undefined' ? process.nextTick
  : (fn) => {
    setTimeout(fn, 0);
  }; */

const re0 = /\$\(.*\)/g;
const cap = '%-cap-%';
const caplen = cap.length;
const re = new RegExp(`(${cap}\\$\\(.*\\))`, 'g');

export function generateTemplate (template) {
  const r = [''];

  template
    .replace(re0, x => cap + x)
    .split(re).forEach(s => {
      if (s.slice(0, caplen) === cap) {
        r.push(s.slice(caplen + 1));
        r.push(Action.of('eval'));
        r.push(Action.of('string'));
        r.push(Action.of('+'));
      } else {
        r.push(String(s));
        r.push(Action.of('+'));
      }
    });
  return r;
}

function lpad (str, n) {
  if (str.length < n) {
    while (str.length < n) {
      str = ` ${str}`;
    }
    return str;
  } else if (str.length > n) {
    return `...${str.slice(3 - n)}`;
  }
  return str;
}

function rtrim (str, n) {
  if (str.length > n) {
    return `${str.slice(0, n - 3)}...`;
  }
  return str;
}

export function formatState ({stack, queue}) {
  stack = lpad(formatArray(stack, 0), 40);
  queue = rtrim(formatArray(queue, 0), 40);
  return `${stack} : ${queue}`;
}

function formatValue (value, depth) {
  if (typeof value === 'symbol') {
    return formatSymbol(value);
  }
  if (Array.isArray(value)) {
    return formatArray(value, depth);
  }
  return inspect(value);
}

function formatSymbol (value) {
  return value.toString().replace(/Symbol\((.*)\)/g, '#$1');
}

function formatArray (obj, depth) {
  obj = obj.map(x => formatValue(x, depth + 1));
  obj = obj.join(' ').replace(/[\s]+/gm, ' ');
  return `[ ${obj} ]`;
}
