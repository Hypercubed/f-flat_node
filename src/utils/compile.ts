import { signature, Any } from '@hypercubed/dynamo';
import is from '@sindresorhus/is';

import { dynamo, Sentence, Word, Seq, Dictionary } from '../types';
import { IIF } from '../constants';

type D = { [key: string]: string };

function create(dictObject: D) {
  class Rewrite {
    @signature()
    array(arr: any[]) {
      return arr.reduce((p, i) => {
        const n = _rewrite(i);
        n instanceof Seq ? p.push(...n.value) : p.push(n);
        return p;
      }, []);
    }
    @signature()
    Sentence(action: Sentence) {
      const expandedValue = _rewrite(action.value);
      return new Sentence(expandedValue, action.displayString);
    }
    @signature()
    Word(action: Word) {
      const path = Dictionary.makePath(action.value).shift();
      const value: string = dictObject[path];
  
      if (is.undefined(value) && (action.value as string)[0] !== IIF)
        return action;
      if (is.function_(value))
        return action;
  
      return new Word(value, action.displayString);
    }
    @signature()
    plainObject(obj: Object) {
      return Object.keys(obj).reduce((p, key) => {
        const n = _rewrite(obj[key]); // todo: think about this, do we ever want to work on anything other than {string: Array}?
        n instanceof Seq
          ? (p[key] = n.value.length === 1 ? n.value[0] : n.value)
          : (p[key] = n);
        return p;
      }, {});
    }
    @signature(Any)
    any(y: any) {
      return y;
    }
  }

  const _rewrite = dynamo.function(Rewrite);
  return _rewrite;
}

export function compile(x: D, y: any) {
  const _rewrite = create(x);
  try {
    return _rewrite(y);
  } catch (e) {
    throw e;
  }
}
