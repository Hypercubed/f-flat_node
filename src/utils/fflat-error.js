import {formatState} from './pprint';

export class FFlatError extends Error {
  constructor (message = 'FFlatError', state = {stack: [], queue: []}) {
    super(message);

    // extending Error is weird and does not propagate `message`
    Reflect.defineProperty(this, 'message', {
      enumerable: false,
      value: message
    });

    Reflect.defineProperty(this, 'name', {
      enumerable: false,
      value: 'FFlatError'
    });

    Reflect.defineProperty(this, 'stack', {
      enumerable: false,
      value:

`${this.name}: ${this.message}
    stack/queue: ${state.stack.length} / ${state.queue.length}
    state:    ${formatState(state)}
`

    });
  }
}
