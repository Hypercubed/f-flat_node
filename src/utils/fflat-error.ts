import { ffPrettyPrint } from './pprint';
import { StackEnv } from '../engine/env';

export class FFlatError extends Error {
  constructor(message = 'FFlatError', state?: StackEnv) {
    super(message);

    // extending Error is weird and does not propagate `message`
    Reflect.defineProperty(this, 'message', {
      enumerable: false,
      value: message
    });

    Reflect.defineProperty(this, 'name', {
      enumerable: false,
      value: 'FFlatError'
    });

    let stackValue: string[];
    if (state) {
      stackValue = [
        `${this.name}: ${this.message}`,
        `stack/queue: ${state.stack.length} / ${state.queue.length}`
      ];

      stackValue.push(`stack trace:`);

      state.trace.forEach(s => {
        stackValue.push(`     ${ffPrettyPrint.formatTrace(s as any)}`);
      });
    } else {
      stackValue = [''];
    }

    Reflect.defineProperty(this, 'stack', {
      enumerable: false,
      value: stackValue.join('\n')
    });
  }
}
