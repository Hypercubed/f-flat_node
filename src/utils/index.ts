export * from './fflat-error';
export * from './logger';
export * from './pprint';
export * from './stringConversion';
export * from './utils';
export * from './kleene-logic';
export * from './json';
export * from './types';
