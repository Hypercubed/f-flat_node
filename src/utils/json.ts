import {  Smykowski, toJSON, encodeSpecialNumbers, encodeUndefined, encodeDates, encodeRegexps, encodeSymbols, encodeSet, encodeMap } from 'smykowski';

const smykowski = new Smykowski()
  .addEncoder(encodeSpecialNumbers)
  .addEncoder(encodeDates)
  .addEncoder(encodeRegexps)
  .addEncoder(encodeSymbols)
  .addEncoder(encodeSet)
  .addEncoder(encodeMap)
  .addEncoder(encodeUndefined)
  .addEncoder(toJSON);

export const encode = smykowski.encode.bind(smykowski);
export const stringifyStrict = smykowski.stringify.bind(smykowski);

