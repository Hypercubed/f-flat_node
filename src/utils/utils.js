import { setIn, freeze, thaw, merge } from 'icepick';
import cloneDeep from 'clone-deep';

import { typed, Action } from '../types/index';

export const arrayRepeat = (a, len) => {
  len = Number(len) | 0;
  let acc = [];
  if (len < 1) {
    return acc;
  }
  let i = -1;
  while (++i < len) {
    acc = acc.concat(a);
  }
  return acc;
};

export const arrayMul = (lhs, rhs) => {
  const len = lhs.length;
  let acc = [];
  if (len < 1) {
    return acc;
  }
  let i = -1;
  while (++i < len) {
    acc = acc.concat(lhs[i], rhs);
  }
  return acc;
};

/* export const listMul = (lhs, rhs) => {
  return lhs.flatMap(x => [x, ...rhs]);
}; */

function baseGet(coll, path) {
  return (path || []).reduce((curr, key) => {
    if (!curr) return;
    curr = Action.isAction(curr) ? curr.value : curr;
    return curr[key];
  }, coll);
}

export const pluck = (context, path) => {
  return baseGet(context, path.split('.'));
}

export const update = (context, path, value) => { // watch immutability
  path = path.split('.');
  if (path.length === 1) {
    return context[path] = value;
  }
  path.reduce((curr, key, currentIndex) => {
    if (currentIndex < path.length - 1) {
      return curr[key];
    }
    curr[key] = value;
  }, context);
}

/* istanbul ignore next */
export function noop() {}

/* istanbul ignore next */
export function throwError(e) {
  throw e;
}

export function isNumeric(n) {
  return !isNaN(parseFloat(n)) && isFinite(n);
}

export function isPromise(val) {
  return val && typeof val.then === 'function';
}

export function isDefined(val) {
  return typeof val !== 'undefined';
}

const __eql = typed('eql', {
  'Array, Array': (a, b) => {
    if (a.length !== b.length) {
      return false;
    }
    for (let i = 0; i < a.length; i++) {
      if (!eql(a[i], b[i])) {
        return false;
      }
    }
    return true;
  },
  'Action, Action': (a, b) => {
    return a.value === b.value;
  },
  'BigNumber | Complex, BigNumber | Complex | number': (a, b) => {
    return a.equals(b);
  },
  'Date, any': (a, b) => {
    return +a === +b;
  },
  'any, Date': (a, b) => {
    return +a === +b;
  },
  'any, any': (a, b) => {
    return a === b;
  }
});

export function eql(a, b) {
  if (a === b || a == b) {
    // eslint-disable-line eqeqeq
    return true;
  }
  return __eql(a, b);
}
