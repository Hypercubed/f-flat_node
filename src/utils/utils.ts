import { assocIn, getIn } from 'icepick';

import { typed, Action } from '../types/index';

export const arrayRepeat = (a: any[], len: any) => {
  len = Number(len) | 0;
  let acc: any[] = [];
  if (len < 1) {
    return acc;
  }
  let i = -1;
  while (++i < len) {
    acc = acc.concat(a);
  }
  return acc;
};

export const arrayMul = (lhs: any[], rhs: any) => {
  const len = lhs.length;
  let acc: any[] = [];
  if (len < 1) {
    return acc;
  }
  let i = -1;
  while (++i < len) {
    acc = acc.concat(lhs[i], rhs);
  }
  return acc;
};

const __eql = typed('deepEquals', {
  'Array, Array': (a: any[], b: any[]): boolean => {
    if (a.length !== b.length) {
      return false;
    }
    for (let i = 0; i < a.length; i++) {
      if (!deepEquals(a[i], b[i])) {
        return false;
      }
    }
    return true;
  },
  'Action, Action': (a: Action, b: Action): boolean => {
    return a.value === b.value;
  },
  'BigNumber | Complex, BigNumber | Complex | number': (a: any, b: any): boolean => {
    return a.equals(b);
  },
  'Date, any': (a: Date, b: any): boolean => {
    return +a === +b;
  },
  'any, Date': (a: any, b: Date): boolean => {
    return +a === +b;
  },
  'any, any': (a: any, b: any): boolean => {
    return a === b;
  }
});

export function deepEquals(a: any, b: any): boolean {
  if (a === b || a == b) { // tslint:disable-line
    // eslint-disable-line eqeqeq
    return true;
  }
  return __eql(a, b);
}

export const toObject = typed('object', {
  Array: (a: any[]) => {
    // hash-map
    const r = {};
    const l = a.length;
    for (let i = 0; l - i > 1; i++) {
      Object.assign(r, { [a[i++]]: a[i] });
    }
    return r;
  },
  any: Object
});