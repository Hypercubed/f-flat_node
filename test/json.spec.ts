import test from 'ava';
import {
  fJSON,
  fValue,
  Decimal,
  Complex,
  Word
} from './setup';
import { stringifyStrict } from '../src/utils/json';

test('prims', t => {
  t.deepEqual(stringifyStrict(42), '42', 'number');
  t.deepEqual(stringifyStrict(-42), '-42', 'negative number');
  t.deepEqual(stringifyStrict('woo!!!'), '"woo!!!"', 'string');
  t.deepEqual(stringifyStrict(true), 'true', 'boolean');
  t.deepEqual(stringifyStrict(null), 'null', 'null');
});

test('special', t => {
  t.deepEqual(stringifyStrict(-0), '{"$numberDecimal":"-0"}', 'negative zero');
  t.deepEqual(stringifyStrict(undefined), '{"$undefined":true}', 'undefined');
  t.deepEqual(stringifyStrict(NaN), '{"$numberDecimal":"NaN"}', 'NaN');
  t.deepEqual(stringifyStrict(Infinity), '{"$numberDecimal":"Infinity"}', 'Infinity');
  t.deepEqual(stringifyStrict(-Infinity), '{"$numberDecimal":"-Infinity"}', 'Infinity');
  t.deepEqual(stringifyStrict(/regexp/img), '{"$regex":"regexp","$options":"gim"}', 'RegExp');
  t.deepEqual(stringifyStrict(new Date(1e12)), '{"$date":"2001-09-09T01:46:40.000Z"}', 'Date');
  t.deepEqual(stringifyStrict(Symbol('a')), '{"$symbol":"a"}', 'Symbol');
});

test('nested', t => {
  testStringify('Array', ['a', 'b', 'c'], '["a","b","c"]');
  testStringify('Array (empty)', [], '[]');
  testStringify('Array (sparse)', [, 'b', , ], '[null,"b",null]');
  testStringify('Object', {foo: 'bar', 'x-y': 'z'}, '{"foo":"bar","x-y":"z"}');
  testStringify('Set', new Set([1, 2, 3]), '{"$set":[1,2,3]}');
  testStringify('Map', new Map([['a', 'b']]), '{"$map":[["a","b"]]}');

  function testStringify(name: string, input: any, expected: string) {
    t.deepEqual(stringifyStrict(input), expected, name);
  }
});

test('nested - special', t => {
  testStringify('Array', ['a', new Date(1e12), 'c'], '["a",{"$date":"2001-09-09T01:46:40.000Z"},"c"]');
  testStringify('Object', {foo: 'bar', 'x-y': new Date(1e12)}, '{"foo":"bar","x-y":{"$date":"2001-09-09T01:46:40.000Z"}}');
  testStringify('Set', new Set([1, new Date(1e12), 3]), '{"$set":[1,{"$date":"2001-09-09T01:46:40.000Z"},3]}');
  testStringify('Map', new Map([['a', new Date(1e12)]]), '{"$map":[["a",{"$date":"2001-09-09T01:46:40.000Z"}]]}');

  function testStringify(name: string, input: any, expected: string) {
    t.deepEqual(stringifyStrict(input), expected, name);
  }
});

test('fflat', t => {
  t.deepEqual(stringifyStrict(new Decimal(42)), '{"$numberDecimal":"42"}', 'decimal');
  t.deepEqual(stringifyStrict(new Complex(42, 3)), '{"@@Complex":{"re":"42","im":"3"}}', 'complex');
  t.deepEqual(stringifyStrict(new Word('word')), '{"@@Action":"word"}', 'word');
});

/* test('cycles', async t => {
  const self = {};
  self.a = self;
  testStringify('Object ( cyclical)', self, '{"a":{"$ref":"#"}}');

  const arr = [];
  arr[0] = arr;
  testStringify('Array (cyclical)', arr, '[{"$ref":"#"}]');

  const map = new Map();
  map.set('self', map);
  testStringify('Map (cyclical)', map, '{"$map":[["self",{"$ref":"#"}]]}');

  const set = new Set();
  set.add(set);
  set.add(42);
  testStringify('Set (cyclical)', set, '{"$set":[{"$ref":"#"},42]}');

  function testStringify(name, input, expected) {
    t.deepEqual(stringifyStrict(input), expected, name);
  }
}); */

test('deep', async t => {
  const obj = {
    string: 'a_string',
    number: 42,
    decimal: new Decimal(42),
    array: ['a_string', 42, new Decimal(42)],
    object: {
      'also a string': 'string',
      'a number': 42
    }
  };

  t.snapshot(stringifyStrict(obj));

  /* obj.object.arr = obj.array;
  t.snapshot(stringifyStrict(obj));

  obj.array.push(obj.decimal);
  t.snapshot(stringifyStrict(obj)); */
});

/* test('references', async t => {
  const a = {a: 1, b: {}};
  a.c = a.b;
  testStringify('Object (references)', a, '{"a":1,"b":{},"c":{"$ref":"#/b"}}');

  const b = {a: 1, b: { c: {} }};
  b.c = b.b.c;
  testStringify('Object (references)', b, '{"a":1,"b":{"c":{}},"c":{"$ref":"#/b/c"}}');

  function testStringify(name, input, expected) {
    t.deepEqual(stringifyStrict(input), expected, name);
  }
}); */

// FF JSON

test('generate json for strings', async t => {
  t.deepEqual((await fJSON('"hello"'))[0], 'hello');
  t.deepEqual((await fJSON('["hello" "world"]'))[0], ['hello', 'world']);
  t.deepEqual((await fJSON('{ x: ["hello" "world"] }'))[0], {
    x: ['hello', 'world']
  });
});

test('generate json for decimals', async t => {
  t.deepEqual((await fJSON('1'))[0], { $numberDecimal: '1' });
  t.deepEqual((await fJSON('[1 2]'))[0], [
    { $numberDecimal: '1' },
    { $numberDecimal: '2' }
  ]);
  t.deepEqual((await fJSON('{ x: [1 2] }'))[0], {
    x: [{ $numberDecimal: '1' }, { $numberDecimal: '2' }]
  });
});

test('generate json for complex values', async t => {
  t.deepEqual((await fJSON('{ x: [1 2], y: i }'))[0], {
    x: [{ $numberDecimal: '1' }, { $numberDecimal: '2' }],
    y: { '@@Complex': { re: '0', im: '1' } }
  });
});

test('generate json for actions', async t => {
  t.deepEqual((await fJSON('"word" :'))[0], { '@@Action': 'word' });
  t.deepEqual((await fJSON('[ "a" "b" ] :'))[0], { '@@Action': ['a', 'b'] });
  t.deepEqual((await fJSON('[ 1 2 ] :'))[0], { '@@Action': [{ $numberDecimal: '1' }, { $numberDecimal: '2' }] });
  t.deepEqual((await fJSON('[ a b ] :'))[0], { '@@Action': [{ '@@Action': 'a' }, { '@@Action': 'b' }] });
});

test('generate json for other values', async t => {  // todo create better json rep
  t.deepEqual((await fJSON('null'))[0], null);
  t.deepEqual((await fJSON('nan'))[0], { $numberDecimal: 'NaN' });
  t.deepEqual((await fJSON('-0'))[0], { $numberDecimal: '-0' });
  t.deepEqual((await fJSON('infinity'))[0], { $numberDecimal: 'Infinity' });
  t.deepEqual((await fJSON('-infinity'))[0], { $numberDecimal: '-Infinity' });
  t.deepEqual((await fJSON('"1/1/1990" date'))[0], { $date: '1990-01-01T07:00:00.000Z' });
  t.deepEqual((await fJSON('"/a./i" regexp'))[0], { $regex: 'a.', $options: 'i' });
  t.deepEqual((await fJSON('#word'))[0], { $symbol: 'word' });
});
