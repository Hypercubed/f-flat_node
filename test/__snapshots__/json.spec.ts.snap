// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deep 1`] = `"{\\"string\\":\\"a_string\\",\\"number\\":42,\\"decimal\\":{\\"$numberDecimal\\":\\"42\\"},\\"array\\":[\\"a_string\\",42,{\\"$numberDecimal\\":\\"42\\"}],\\"object\\":{\\"also a string\\":\\"string\\",\\"a number\\":42}}"`;
